rust-runtime/smithy-json/fuzz/fuzz_targets/common.rs

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+use serde_json::{Map, Value};
+use smithy_json::deserialize::{Error, Token};
+use smithy_types::Number;
+
+pub fn run_data(data: &[u8]) {
+    // Parse through with smithy-json first to make sure it doesn't panic on invalid inputs
+    if let Ok(tokens) =
+        smithy_json::deserialize::json_token_iter(data).collect::<Result<Vec<Token>, Error>>()
+    {
+        // Exercise string unescaping since the later comparison against Serde
+        // reserializes, and thus, loses UTF-16 surrogate pairs.
+        for token in tokens {
+            if let Token::ValueString(escaped) = token {
+                if let Ok(unescaped) = escaped.to_unescaped() {
+                    let serde_equiv = serde_json::from_str::<String>(&format!(
+                        "\"{}\"",
+                        escaped.as_escaped_str()
+                    ))
+                    .unwrap();
+                    assert_eq!(serde_equiv, unescaped);
+                }
+            }
+        }
+    }
+
+    // Now parse with Serde, and if it's valid, compare the two and panic if different
+    let serde_value = serde_json::from_slice::<Value>(data);
+    if let Ok(value) = serde_value {
+        // Re-serialize to normalize the large floating point numbers
+        let json = serde_json::to_string(&value).unwrap();
+
+        let tokens = smithy_json::deserialize::json_token_iter(json.as_bytes())
+            .collect::<Result<Vec<Token>, Error>>()
+            .unwrap();
+        let mut token_iter = RewindingTokenIter::new(tokens);
+        let converted_value = convert_tokens(&mut token_iter);
+        assert_eq!(None, token_iter.next());
+        assert_eq!(value, converted_value);
+    }
+}
+
+/// Utility for iterating over the tokens while being able to rewind
+struct RewindingTokenIter<'a> {
+    tokens: Vec<Token<'a>>,
+    position: usize,
+}
+
+impl<'a> RewindingTokenIter<'a> {
```

Comment:
doesn't really matter, but this could be replaced with https://doc.rust-lang.org/std/iter/struct.Peekable.html

Reply:
I wasn't aware of this. Thanks.